Docs: Add queries.md (#24363)
grafana/grafana: docs/sources/panels/queries.md

In addition to the data sources that you have configured in your Grafana, there are three special data sources available:

* **Grafana -** A built-in data source that generates random walk data. Useful for testing visualizations and running experiments.
* **Mixed -** Select this to query multiple data sources in the same panel. When this data source is selected, Grafana allows you to select a data source for every new query that you add.
  * The first query will use the data source that was selected before you selected **Mixed**.
  * You cannot change an existing query to use the Mixed Data Source.
* **Dashboard -** Select this to use a result set from another panel in the same dashboard.

### Data source options

Click **Options** next to the data source selector to see settings for your selected data source. Changes you make here affect only this panel.

{{< docs-imagebox img="/img/docs/queries/data-source-options-7-0.png" class="docs-image--no-shadow" max-width="400px" >}}

Grafana sets defaults that are shown in dark gray text. Changes are displayed in white text. To return a field to the default setting, delete the white text from the field.

Panel data source options:

> 

Author: @torkelo or @dprokop , please enter definitions for these fields.

Author: @daniellee , @torkelo ?

Reviewer: ### max data points

If the data source supports it, sets the maximum numbers of data points for each series returned. If the query returns more data points than the `max data points` setting then the data source will consolidate them (reduce the number of points returned by aggregating them together by average or max or other function). There are two main reasons for limiting the number of points, performance and smoothing the line. The default value is the width (or number of pixels) of the graph as there is no point in having more data points than the graph panel can display.

With streaming data, the max data points value is used for the rolling buffer. (Streaming is a continuous flow of data and buffering is a way of dividing the stream into chunks). Loki streams data in the live tailing mode.

This article describes how they can be confusing - the default consolidation function for Graphite is the average when using maxdatapoints and this can be surprising if you are expecting something else: https://addshore.com/2018/09/grafana-graphite-and-maxdatapoints-confusion-for-totals/

Reviewer: ### interval

The interval is a time span can be used when aggregating or grouping by time. Grafana automatically calculates an appropriate interval and it can be used as a variable in templated queries. The variable is either in seconds: `$__interval` or in milliseconds: `$__interval_ms`. It is typically used in aggregation functions like sum or average. For example, a Prometheus query using the interval variable: `rate(http_requests_total[$__interval])`.

This automatic interval is calculated based on the width of the graph. If the user zooms out a lot then the interval becomes greater, resulting in a more coarse grained aggregation whereas if the use zooms in then the interval decreases resulting in a more fine grained aggregation.

More details here: https://github.com/grafana/grafana/blob/add-queries.md/docs/sources/variables/global-variables.md

Reviewer: ### Min interval

Sets a minimum limit for the automatically calculated interval. Typically set to the minimum scrape interval (using Prometheus terminology here). If a data point is saved every 15 seconds then there is no point in having an interval that is lower than that. Another use case is set it to a higher minimum than the scrape interval to get more coarse grained, performant queries.

Reviewer: ### Relative time

You can override the relative time range for individual panels, causing them to be different than what is selected in the Dashboard time picker in the upper right. This allows you to show metrics from different time periods or days on the same dashboard.

Reviewer: ### Time shift

The time shift function is another way to override the time range for individual panels. It only works with relative time ranges and allows you to adjust the time range. For example, shifting the time range for the panel to be 2 hours earlier than the dashboard time picker.

https://grafana.com/docs/grafana/latest/reference/timerange/#panel-time-overrides-and-timeshift

Reviewer: ### Cache timeout

If your time series store has a query cache, this option can override the default cache timeout. Specified as a numeric value in seconds. 

Here is the original issue. The use case was to override the Graphite cache time of 1m to be able to more up to date data.

https://github.com/grafana/grafana/issues/266

+++
title = "Queries"
type = "docs"
[menu.docs]
identifier = "queries"
parent = "panels"
weight = 300
+++

# Queries

> **Note:** This documentation refers to a feature only available in Grafana 7.0 beta.

> _Queries_ are how Grafana panels communicate with data sources to get data for the visualization. A query is basically a question written in the query language used by the data source. Grafana asks, "Hey data source, would you send me this data, organized this way?" If the query is properly formed, than the data source responds. How often the query is sent to the data source and how many data points are collected can be adjusted in the panel data source options.
> 

Reviewer 2: ```suggestion
_Queries_ are how Grafana panels communicate with data sources to get data for the visualization. A query is a question written in the query language used by the data source. Grafana asks, "Hey data source, would you send me this data, organized this way?" If the query is properly formed, then the data source responds. How often the query is sent to the data source and how many data points are collected can be adjusted in the panel data source options.

```

Basically removed the "basically" and fixed a typo... :D

Grafana supports up to 26 queries per panel.

## Query editors

> Query editors are forms that help you write queries. Depending on your data source, the query editor might completion, metric names, or variable suggestion.

Reviewer 2: `might completion`? this feels weird

Author: Yeah, that sort of thing is what happens when I try to write documentation when my brain is fried. This is why we have other people edit. :)

Author: Fixed. Hopefully. Please make sure what I wrote makes sense.

Reviewer 2: all good now!

Because of the difference between query languages, data sources may have query editors that look different. Here are two examples of query editors:

**InfluxDB query editor**

{{< docs-imagebox img="/img/docs/queries/influxdb-query-editor-7-0.png" class="docs-image--no-shadow" max-width="1000px" >}}

**Prometheus (PromQL) query editor**

{{< docs-imagebox img="/img/docs/queries/prometheus-query-editor-7-0.png" class="docs-image--no-shadow" max-width="1000px" >}}

## Query syntax

> Every data source has a different query language and syntax to ask for the data. Here are two query examples:

Reviewer 2: ```suggestion
Data sources have different query languages and syntaxes to ask for the data. Here are two query examples:
```

SQL datasource have quite similar syntax. This form sounds better to me.

{{< docs-imagebox img="/img/docs/queries/query-editor-7-0.png" class="docs-image--no-shadow" max-width="1000px" >}}

### Data source selector

The data source selector is a drop-down list. Click it to select a data source you have added. When you create a panel, Grafana automatically selects your default data source. For more information about adding data sources, refer to [Add a data source]({{< relref "../features/datasources/add-a-data-source.md" >}}).

{{< docs-imagebox img="/img/docs/queries/data-source-selector-7-0.png" class="docs-image--no-shadow" max-width="250px" >}}

In addition to the data sources that you have configured in your Grafana, there are three special data sources available:

* **Grafana -** A built-in data source that generates random walk data. Useful for testing visualizations and running experiments.
* **Mixed -** Select this to query multiple data sources in the same panel. When this data source is selected, Grafana allows you to select a data source for every new query that you add.
  * The first query will use the data source that was selected before you selected **Mixed**.
  * You cannot change an existing query to use the Mixed Data Source.

> * **Dashboard -** Select this to query a result set from another panel in the same dashboard.

Reviewer 2: ```suggestion
* **Dashboard -** Select this to use a result set from another panel in the same dashboard.
```

### Data source options

> Click **Options** next to the data source selector to see settings for your selected data source. Changes you make here affect only this panel.

Reviewer 3: I would not name these data source options but more like query options (that impact all queries)

Reviewer 2: yes, agree

Author: They are next to the data source and they are options for how often you query the data source and how many points you get back from it. Some fields are identical to the data source settings. Based on the current UI, they are clearly data source options.

Reviewer 3: no, they are not. they are query options, they are options like how many data points should queries return, what should the time range for the queries be, what is the cache timeout for the queries, etc. So they are not options for the data source, but options for queries listed below 

Data source options are on a completely different page. It's true that min interval can ALSO be set on data source level, but that is as I said in a different view. 

If you switch data source for example these options are NOT cleared as they are query options for this panel, not linked to the data source (except some options are linked to what data source type, only some options show the cache timeout option for example)

Author: That might be what they are, but the UI is telling me (the user) that Options are related to the data sources, because they are right next to the data source selected. @dprokop, @sarlinska , any thoughts on this UI vs. functionality puzzle?

Author: ![image](https://user-images.githubusercontent.com/52059945/81314808-624ff200-903e-11ea-8d50-4592509d4149.png)

Author: Do you want to rename "Options" to "Query options"? Because right now, "Options next to the Data source selector field" is what this is.

Reviewer 3: Yes, these should be named query options, I think they where I at some point :)

Author: So can we change it back for 7.0? It looks like we have the space, and it would clarify for the user what they are looking at. (And make it easier for me to document.)

Reviewer 2: @oddlittlebird i'll update it

Reviewer 2: fixed in v

Click **Query inspector** to open the Query tab of the panel inspector where you can see the query request sent by the panel and the response. 

Click **Refresh** to see the full text of the request sent by this panel to the server.

> **Note:** You need to add at least one query before the Query inspector can return results.

For more information about the panel inspector, refer to [Inspect a panel]({{< relref "inspect-panel.md" >}}).

### Query editor list

In the UI, queries are organized in collapsible query rows. Each query row contains a query editor and is identified with a letter (A, B, C, and so on). 

You can:

* **Change the order of queries** by moving up [INLINE ICON] or down [INLINE ICON].
  The order of results reflects the order of the queries, so you can often adjust your visual results based on query order.

* **Copy a query** [INLINE ICON]. 
  Duplicating queries is useful when working with multiple complex queries that are similar and you want to either experiment with different variants or do minor alterations.  

* **Hide a query** [INLINE ICON]. 
  Grafana does not send hidden queries to the data source.

* **Remove a query** [INLINE ICON]

>   Removing a query permanently deletes it, but sometimes you can recover deleted queries by reverting to previously saved versions of the panel.

Reviewer 2: @oddlittlebird added the inline icon images. I have understood you want the query editor images as well so I have updated ALL the query images

Author: Thanks, I'll take a look after the All Hands.
